internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package structuredmerge
+
+import (
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+const topologyManagerName = "topology"
+
+type serverSidePatchHelper struct {
+	client         client.Client
+	modified       *unstructured.Unstructured
+	hasChanges     bool
+	hasSpecChanges bool
+}
+
+// NewServerSidePatchHelper returns a new PatchHelper using server side apply.
+func NewServerSidePatchHelper(scheme *runtime.Scheme, original, modified client.Object, c client.Client, opts ...HelperOption) (PatchHelper, error) {
+	helperOptions := &HelperOptions{}
+	helperOptions = helperOptions.ApplyOptions(opts)
+	helperOptions.allowedPaths = []contract.Path{
+		// apiVersion, kind, name and namespace are required field for a server side apply intent.
+		{"apiVersion"},
+		{"kind"},
+		{"metadata", "name"},
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"metadata", "ownerReferences"},
+		{"spec"},
+	}
+
+	// If required, convert the original and modified objects to unstructured and filter out all the information
+	// not relevant for the topology controller.
+
+	var originalUnstructured *unstructured.Unstructured
+	if original != nil {
+		originalUnstructured = &unstructured.Unstructured{}
+		switch original.(type) {
+		case *unstructured.Unstructured:
+			originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
```

Comment:
There is a potential case of panic here. For context: I did detect this when I tried to simplify the code to not need to distinct between create and update of objects.

```suggestion
			if reflect.ValueOf(original).IsValid() && !reflect.ValueOf(original).IsNil() {
				originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
			}
```

Same below for modified (may not be necessary because we don't handle deletion but would prevent panic's.


Test case for `TestServerSideApply`:

```go
	t.Run("Server side apply detect changes on object creation", func(t *testing.T) {
		g := NewWithT(t)

		var original *unstructured.Unstructured
		modified := obj.DeepCopy()

		p0, err := NewServerSidePatchHelper(original, modified, env.GetClient())
		g.Expect(err).ToNot(HaveOccurred())
		g.Expect(p0.HasChanges()).To(BeTrue())
		g.Expect(p0.HasSpecChanges()).To(BeTrue())
	})
```

Reply:
maybe it would be also fair to return an error in such cases.

Reply:
Another edge-case:
```go
	t.Run("Server side apply detect changes on object creation passing nil", func(t *testing.T) {
		g := NewWithT(t)

		var original *clusterv1.MachineDeployment
		modified := obj.DeepCopy()

		p0, err := NewServerSidePatchHelper(original, modified, env.GetClient())
		g.Expect(err).ToNot(HaveOccurred())
		g.Expect(p0.HasChanges()).To(BeTrue())
		g.Expect(p0.HasSpecChanges()).To(BeTrue())
	})
```

Possible solution instead of above proposed:

```go
	var originalUnstructured *unstructured.Unstructured
	if original != nil {
		originalUnstructured = &unstructured.Unstructured{}
		if reflect.ValueOf(original).IsValid() && !reflect.ValueOf(original).IsNil() {
			switch original.(type) {
			case *unstructured.Unstructured:
				originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
			default:
				if err := c.Scheme().Convert(original, originalUnstructured, nil); err != nil {
					return nil, errors.Wrap(err, "failed to convert original object to Unstructured")
				}
			}
			filterObject(originalUnstructured, helperOptions)
		}
	}
```

Reply:
fixed